### DISCUSSION TITLE
Using SCons to install development dependencies

### DISCUSSION THREAD

**Discussion by abitrolly** [SCons/scons#4269] [Q&A]
I like idempotent nature of SCons for building files. If source files are changed then rebuild target files, but if source files are intact, do nothing.

I see how the same principle can be applied to installing packages. If some package version is not installed - find the best way and install it, but if the package is already installed, do nothing.

Can SCons be used for this?

**Comment by mwichmann** [comment]
Could you clarify this a bit?  What kind of packages?  There's an external project which adds some support for Python packages, but more for construction than for installation (https://github.com/dholth/enscons).  For systems that have "package managers" I'd say they already do this - they install only if necessary, unless you give extra arguments to force the issue. Windows doesn't come with the same kind of package manager, NuGet provides some of that capability, and Chocolatey adds more infrastructure on top of it, so you can get the same kind of effect.

**Reply by abitrolly** [reply]
@mwichmann I changed the title to clarify that "packages" are development dependencies. They could be binaries installed with `brew`, `asdf`, system level package managers, coming as a container images or compiled from source (with distributed caching of reproducible builds signed by multiple parties).

Take this issue for the context - https://gitlab.com/gitlab-org/gitlab-development-kit/-/issues/1585 - for the GitLab development, almost all package sources I mentioned are used, and that comes with significant drawbacks. Like installing specific version of PostgreSQL with `asdf` takes about 30 minutes, because PostgreSQL ships with no binaries for Linux. If that specific version is already available from OS repository, it makes sense to use that, but `asdf` can not do this.

I just thought that detecting which package dependency is needed on specific platform, which is the best source to get that dependency, and tracking if the dependency is already installed (built) to avoid repeating extra work - that functionality fits well into SCons processing model. Am I right?
